File: src/cowtree/submodules.py
"""Admit pinned dependencies and verify private materializations without updating their refs.

Submodule contents become read-only source in the managed Git projection.
Their original Git control directories are never copied into snapshots or leaves.
"""

from dataclasses import replace
import errno
import os
from pathlib import Path
import tempfile
import unicodedata

from pydantic import TypeAdapter

from cowtree.errors import CowtreeError, CowtreeErrorCode
from cowtree.git import GitRepository
from cowtree.install import local_path
from cowtree.metadata_types import Entry, ReadOnlyKind
from cowtree.submodule_types import PinnedSubmodule
from cowtree.tree_types import PathPolicy
from cowtree.types import FileMode, TrackedFile


PIN_FILE = ".cowtree-pin"
PIN = TypeAdapter(PinnedSubmodule)


def admit(repository: GitRepository, policy: PathPolicy) -> PathPolicy:
    """Pin clean initialized children before reserving a workspace directory."""
    if policy.pins:
        raise CowtreeError(CowtreeErrorCode.INVALID_ARGUMENTS, "pins are derived from Git")
    checkout = repository.snapshot(submodules=policy.submodules)
    for pin in checkout.submodules:
        path = local_path(root=repository.path, relative=pin.path)
        if path.is_symlink() or not path.is_dir():
            raise CowtreeError(
                CowtreeErrorCode.SUBMODULE_UNSUPPORTED, f"uninitialized submodule: {pin.path}"
            )
        child = GitRepository.discover(io=repository.io, source=path)
        if child.path != path.resolve():
            raise CowtreeError(
                CowtreeErrorCode.SUBMODULE_UNSUPPORTED, f"uninitialized submodule: {pin.path}"
            )
        snapshot = child.snapshot()
        validate_links(child.path, snapshot.files)
        if snapshot.commit != pin.commit:
            raise CowtreeError(
                CowtreeErrorCode.HEAD_MISMATCH, f"submodule differs from gitlink: {pin.path}"
            )
        verify_tree(child, path, pin.commit)
        if child.capture(args=["ls-files", "--others", "--directory", "-z"]):
            raise CowtreeError(
                CowtreeErrorCode.DIRTY_SOURCE, f"untracked submodule contents: {pin.path}"
            )
        if os.path.lexists(path / PIN_FILE):
            raise CowtreeError(
                CowtreeErrorCode.INVALID_ARGUMENTS, f"reserved dependency marker: {pin.path}"
            )
        for prefix in (*policy.derived, *policy.ephemeral, *policy.ignored):
            if overlaps(pin.path, prefix):
                raise CowtreeError(
                    CowtreeErrorCode.INVALID_ARGUMENTS,
                    f"dependency conflicts with path policy: {pin.path}",
                )
    result = replace(policy, pins=checkout.submodules)
    return result


def overlaps(left: str, right: str) -> bool:
    result = left == right or left.startswith(right + "/") or right.startswith(left + "/")
    return result


def validate_links(root: Path, files: tuple[TrackedFile, ...]) -> None:
    """Copied dependency links cannot retain references to the source or outside files."""
    for entry in files:
        if entry.mode is not FileMode.SYMLINK:
            continue
        path = local_path(root, entry.path)
        link = Path(os.readlink(path))
        logical = Path(os.path.normpath(path.parent / link))
        try:
            resolved = path.resolve()
        except (RuntimeError, OSError) as error:
            if isinstance(error, OSError) and error.errno != errno.ELOOP:
                raise
            raise CowtreeError(
                CowtreeErrorCode.SUBMODULE_UNSUPPORTED, f"cyclic dependency link: {path}"
            ) from error
        if (
            link.is_absolute()
            or not logical.is_relative_to(root)
            or not resolved.is_relative_to(root)
        ):
            raise CowtreeError(
                CowtreeErrorCode.SUBMODULE_UNSUPPORTED, f"dependency link escapes capture: {path}"
            )


def materialize(repository: GitRepository, source: Path, target: Path, policy: PathPolicy) -> None:
    """Verify copied dependency bytes against fresh Git indexes before recording their pins."""
    for pin in policy.pins:
        child = GitRepository.discover(io=repository.io, source=source / pin.path)
        if child.head() != pin.commit:
            raise CowtreeError(
                CowtreeErrorCode.HEAD_MISMATCH, f"submodule HEAD changed: {pin.path}"
            )
        copied = local_path(root=target, relative=pin.path)
        verify_tree(child, copied, pin.commit)
        with (copied / PIN_FILE).open("xb") as stream:
            stream.write(PIN.dump_json(pin))


def verify_tree(repository: GitRepository, tree: Path, commit: str) -> None:
    """Use an owned stat-empty index so caller index caches cannot hide changed bytes."""
    directory = repository.capture(
        args=["rev-parse", "--path-format=absolute", "--git-dir"]
    ).removesuffix("\n")
    command = [
        "git",
        "--git-dir",
        directory,
        "--work-tree",
        str(tree),
        "-c",
        "core.filemode=true",
        "-c",
        "core.fsmonitor=false",
        "-c",
        "core.ignorestat=false",
    ]
    with tempfile.TemporaryDirectory(prefix="cowtree-pin-index-") as scratch:
        environment = {**os.environ, "GIT_INDEX_FILE": str(Path(scratch) / "index")}
        repository.io.run(argv=[*command, "read-tree", commit], env=environment)
        result = repository.io.run(
            argv=[*command, "status", "--porcelain=v1", "-z", "--untracked-files=no"],
            env=environment,
        )
    if result.stdout:
        raise CowtreeError(
            CowtreeErrorCode.DIRTY_SOURCE, f"dependency differs from pinned tree: {tree}"
        )


def scope(path: str, policy: PathPolicy) -> str | None:
    for pin in policy.pins:
        prefix = unicodedata.normalize("NFC", pin.path)
        if path == prefix or path.startswith(prefix + "/"):
            return prefix
    return None


def verify_manifest(actual: dict[str, Entry], expected: dict[str, Entry]) -> None:
    """A checkpoint cannot change, remove, or extend an inherited dependency namespace."""
    for path in actual.keys() | expected.keys():
        before = expected.get(path)
        after = actual.get(path)
        protected = any(
            entry is not None and isinstance(entry.kind, ReadOnlyKind) for entry in (before, after)
        )
        if protected and before != after:
            raise CowtreeError(
                CowtreeErrorCode.DEPENDENCY_CHANGED, f"read-only dependency changed: {path}"
            )

File: crates/workspace/src/submodules.rs
// Copyright (c) 2026 Windsor Nguyen

//! Admit initialized pinned dependencies and verify their copied bytes with fresh indexes.

use std::{
    fs,
    path::{Path, PathBuf},
};

use crate::{
    CommitId, Error, Policy, Result, SubmodulePolicy, error::Issue, git::Repository, paths,
    records, types::Pin,
};

const MARKER: &str = ".cowtree-pin";

impl Repository {
    pub(crate) fn snapshot(&self, policy: SubmodulePolicy) -> Result<(CommitId, Vec<Pin>)> {
        let sparse =
            self.output(self.command()?.args(["config", "--bool", "core.sparseCheckout"]))?;
        if sparse.stdout == b"true\n" {
            return Err(Issue::SparseCheckout.at(&self.root));
        }
        if !matches!(sparse.status.code(), Some(0 | 1)) {
            return Err(cowtree_git::Error::Command {
                status: sparse.status,
                stderr: String::from_utf8_lossy(&sparse.stderr).into_owned(),
            }
            .into());
        }
        let commit = self.head()?;
        let tree = self.capture(self.command()?.args(["ls-tree", "-r", "-z", commit.as_str()]))?;
        let mut pins = Vec::new();
        for record in tree.split(|byte| *byte == 0).filter(|record| !record.is_empty()) {
            let record = paths::utf8(record, &self.root)?;
            let (header, path) =
                record.split_once('\t').ok_or_else(|| Issue::IncompleteRecord.at(&self.root))?;
            let fields: Vec<_> = header.split(' ').collect();
            if fields.len() != 3 {
                return Err(Issue::IncompleteRecord.at(&self.root));
            }
            if fields[0] == "160000" {
                if policy == SubmodulePolicy::Reject {
                    return Err(Issue::InvalidSubmodule.at(&self.root.join(path)));
                }
                pins.push(Pin { path: path.into(), commit: CommitId::parse(fields[2].into())? });
            } else if !matches!(fields[0], "100644" | "100755" | "120000") {
                return Err(Issue::UnsupportedEntry.at(&self.root.join(path)));
            }
        }
        let flags = self.capture(self.command()?.args(["ls-files", "-v", "-z"]))?;
        if flags.split(|byte| *byte == 0).any(|record| {
            record.first().is_some_and(|flag| flag.is_ascii_lowercase() || *flag == b'S')
        }) || !self.status()?.is_empty()
        {
            return Err(Issue::DirtyCheckout.at(&self.root));
        }
        Ok((commit, pins))
    }

    pub(crate) fn verify_tree(&self, tree: &Path, commit: &CommitId) -> Result<()> {
        let git_directory =
            self.text(self.command()?.args(["rev-parse", "--path-format=absolute", "--git-dir"]))?;
        let temporary = tempfile::Builder::new()
            .prefix("cowtree-pin-index-")
            .tempdir()
            .map_err(|error| Error::io(tree, error))?;
        let command = || -> Result<std::process::Command> {
            let mut command = self.command()?;
            command
                .args(["--git-dir", &git_directory, "--work-tree"])
                .arg(tree)
                .args([
                    "-c",
                    "core.filemode=true",
                    "-c",
                    "core.fsmonitor=false",
                    "-c",
                    "core.ignorestat=false",
                ])
                .env("GIT_INDEX_FILE", temporary.path().join("index"));
            Ok(command)
        };
        self.capture(command()?.args(["read-tree", commit.as_str()]))?;
        let changed = self.capture(command()?.args([
            "status",
            "--porcelain=v1",
            "-z",
            "--untracked-files=no",
        ]))?;
        if !changed.is_empty() {
            return Err(Issue::DirtyCheckout.at(tree));
        }
        Ok(())
    }
}

pub(crate) fn admit(repository: &Repository, policy: &Policy) -> Result<Policy> {
    if !policy.pins.is_empty() {
        return Err(Issue::InvalidSubmodule.at(&repository.root));
    }
    let (_, pins) = repository.snapshot(policy.submodules)?;
    for pin in &pins {
        let path = paths::local(&repository.root, &pin.path)?;
        let metadata = fs::symlink_metadata(&path).map_err(|error| Error::io(&path, error))?;
        if !metadata.is_dir() {
            return Err(Issue::InvalidSubmodule.at(&path));
        }
        let child = Repository::discover(&path)?;
        if child.root != path {
            return Err(Issue::InvalidSubmodule.at(&path));
        }
        let (commit, _) = child.snapshot(SubmodulePolicy::Reject)?;
        if commit != pin.commit {
            return Err(Issue::HeadChanged.at(&path));
        }
        validate_links(&child)?;
        child.verify_tree(&path, &pin.commit)?;
        if !child
            .capture(child.command()?.args(["ls-files", "--others", "--directory", "-z"]))?
            .is_empty()
        {
            return Err(Issue::DirtyCheckout.at(&path));
        }
        if paths::exists(&path.join(MARKER))? {
            return Err(Issue::InvalidSubmodule.at(&path));
        }
        for prefix in policy.derived.iter().chain(&policy.ephemeral).chain(&policy.ignored) {
            if overlaps(&pin.path, prefix) {
                return Err(Issue::ConflictingPolicy.at(&path));
            }
        }
    }
    let mut admitted = policy.clone();
    admitted.pins = pins;
    Ok(admitted)
}

pub(crate) fn materialize(
    repository: &Repository,
    source: &Path,
    target: &Path,
    policy: &Policy,
) -> Result<()> {
    for pin in &policy.pins {
        let mut child = Repository::discover(&source.join(&pin.path))?;
        child.locks = repository.locks.clone();
        if child.head()? != pin.commit {
            return Err(Issue::HeadChanged.at(&child.root));
        }
        let copied = paths::local(target, &pin.path)?;
        child.verify_tree(&copied, &pin.commit)?;
        let marker = copied.join(MARKER);
        let file = fs::File::options()
            .write(true)
            .create_new(true)
            .open(&marker)
            .map_err(|error| Error::io(&marker, error))?;
        serde_json::to_writer(&file, pin).map_err(|error| Error::record(&marker, error))?;
        records::sync_file(&file).map_err(|error| Error::io(&marker, error))?;
    }
    Ok(())
}

fn overlaps(left: &str, right: &str) -> bool {
    left == right
        || left.strip_prefix(right).is_some_and(|tail| tail.starts_with('/'))
        || right.strip_prefix(left).is_some_and(|tail| tail.starts_with('/'))
}

fn validate_links(repository: &Repository) -> Result<()> {
    let tree = repository.capture(repository.command()?.args(["ls-tree", "-r", "-z", "HEAD"]))?;
    for record in tree.split(|byte| *byte == 0).filter(|record| record.starts_with(b"120000 ")) {
        let (_, name) = record.split_at(
            record
                .iter()
                .position(|byte| *byte == b'\t')
                .ok_or_else(|| Issue::IncompleteRecord.at(&repository.root))?
                + 1,
        );
        let path = paths::local(&repository.root, &paths::utf8(name, &repository.root)?)?;
        let link = fs::read_link(&path).map_err(|error| Error::io(&path, error))?;
        if link.is_absolute() {
            return Err(Issue::InvalidSubmodule.at(&path));
        }
        let mut logical = path.parent().ok_or_else(|| Issue::InvalidPath.at(&path))?.to_path_buf();
        for component in link.components() {
            match component {
                std::path::Component::ParentDir => {
                    logical.pop();
                }
                std::path::Component::CurDir => {}
                std::path::Component::Normal(part) => logical.push(part),
                _ => return Err(Issue::InvalidSubmodule.at(&path)),
            }
        }
        let resolved: PathBuf = resolve_links(&path)?;
        if !logical.starts_with(&repository.root) || !resolved.starts_with(&repository.root) {
            return Err(Issue::InvalidSubmodule.at(&path));
        }
    }
    Ok(())
}

/// Resolve links iteratively, allowing missing terminal names inside the dependency.
fn resolve_links(path: &Path) -> Result<PathBuf> {
    use std::{
        collections::{BTreeSet, VecDeque},
        ffi::OsString,
    };
    enum Step {
        Component(OsString),
        EndLink(PathBuf),
    }
    let mut pending: VecDeque<_> =
        path.components().map(|part| Step::Component(part.as_os_str().to_owned())).collect();
    let mut resolved = PathBuf::new();
    let mut active = BTreeSet::new();
    while let Some(step) = pending.pop_front() {
        let part = match step {
            Step::EndLink(path) => {
                active.remove(&path);
                continue;
            }
            Step::Component(part) => part,
        };
        if part == "/" {
            resolved = PathBuf::from("/");
            continue;
        }
        if part == "." {
            continue;
        }
        if part == ".." {
            resolved.pop();
            continue;
        }
        let candidate = resolved.join(&part);
        match fs::symlink_metadata(&candidate) {
            Ok(metadata) if metadata.is_symlink() => {
                if !active.insert(candidate.clone()) {
                    return Err(Issue::InvalidSubmodule.at(path));
                }
                let link =
                    fs::read_link(&candidate).map_err(|error| Error::io(&candidate, error))?;
                pending.push_front(Step::EndLink(candidate));
                for component in link.components().rev() {
                    pending.push_front(Step::Component(component.as_os_str().to_owned()));
                }
            }
            Ok(_) => resolved.push(part),
            Err(error)
                if matches!(
                    error.kind(),
                    std::io::ErrorKind::NotFound | std::io::ErrorKind::NotADirectory
                ) =>
            {
                resolved.push(part)
            }
            Err(error) => return Err(Error::io(&candidate, error)),
        }
    }
    Ok(resolved)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn link_resolution_distinguishes_cycles_from_repeated_noncyclic_paths() -> Result<()> {
        let root = tempfile::tempdir().map_err(|error| Error::io(Path::new("temporary"), error))?;
        let root = root.path().canonicalize().map_err(|error| Error::io(root.path(), error))?;
        std::os::unix::fs::symlink(".", root.join("link"))
            .map_err(|error| Error::io(&root, error))?;
        assert_eq!(resolve_links(&root.join("link/link/future"))?, root.join("future"));
        std::os::unix::fs::symlink("cycle", root.join("cycle"))
            .map_err(|error| Error::io(&root, error))?;
        assert!(matches!(
            resolve_links(&root.join("cycle")),
            Err(Error::State { issue: Issue::InvalidSubmodule, .. })
        ));
        Ok(())
    }
}
